test/collection_serializer_test.rb

```diff
@@ -23,6 +23,21 @@ def build_named_collection(*resource)
         resource
       end
 
+      def build_named_collection_with_matching_serializer(*resource)
+        serialized_resource = Class.new(::Model)
+        serializer = Class.new(ActiveModel::Serializer) do
+          def json_key
+            'resources'
+          end
+        end
+
+        Object.const_set(:SerializedResource, serialized_resource)
+        Object.const_set(:SerializedResourceSerializer, serializer)
+
+        resource.define_singleton_method(:name) { 'SerializedResource' }
```

Comment:
1. I'd rather this be in a block so that we can also teardown the classes after defining them
1. I think we can consolidate the code here a little

```ruby
def with_serializable_named_model(name)
  model = Class.new(::Model) do
    def self.name
      'SerializableResource'
    end
  end
  Object.const_set(:SerializableResource, model)
  serializer = Class.new(ActiveModel::Serializer) do
    def json_key
      name
    end
  end
  Object.const_set(:SerializableResourceSerializer, serializer)
  yield model
ensure
  Object.send(:remove_const, :SerializableResource)
  Object.send(:remove_const, :SerializableResourceSerializer)
end
```

then

```ruby
def test_json_key_with_resource_with_name_and_serializer
  with_serializable_named_collection('resources') do |model|
     serializer_instance = collection_serializer.new([model])
     assert_equal 'resources', serializer_instance.json_key
  end
end
```